aten/src/ATen/core/dynamic_type.h

```diff
@@ -0,0 +1,156 @@
+  _(AnyList, (kDynamicListTypeBit | kDynamicAnyTypeBit))                     \
+  _(AnyTuple,                                                                \
+    (kDynamicTupleTypeBit | kDynamicCovariantTypeBit | kDynamicAnyTypeBit))  \
+  _(Any, 0xffffffff)
+
+namespace c10 {
+
+struct LabeledDynamicType;
+class DynamicType;
+using DynamicTypePtr = std::shared_ptr<DynamicType>;
+
+/**
+ * DynamicType is designed as a low dependency type system for TorchScript. The
+ * existing JIT types are used for both compilation and runtime, which makes
+ * sense for server contexts because we often compile and run the model in
+ * the same process, however this doesn't hold for mobile devices where we
+ * always compiles a model ahead of time, therefore there will be dependencies
+ * which are not needed, but built with mobile runtime causing binary size
+ * bloat, by design. Why linkers don't strip unused deps? It is because the
+ * existing (server) JIT types are defined by overriding virtual functions from
+ * the JIT base type, which makes controlling the binary size for mobile even
+ * harder because linkers have to include all possible overrides for a single
+ * callsite.
+ *
+ * The core problem is about the complexity to implement and maintain a single
+ * type system for both analysis and execution purposes. Although they should
+ * have the exactly same semantics, in practice implement a unified abstraction
+ * adds conceptual and representational overhead for both sides of the world.
+ *
+ * To address the issues, DynamicType implements a minimal subset of JIT types
+ * and uses a generic algorithm to test all subtyping relations. To achieve
+ * this, we assign each dynamic type a single integer tag to represent its
+ * semantics. More specifically, a dynamic type is defined as a set of "control
+ * bits" and "data bits", where control bits describe the special behavior when
+ * testing a type and data bits map to identity of each nominal type. We use bit
+ * operations to perform all the tests.
+ *
+ * For example, a "covariant bit" is a control bit used to describe if a type
+ * is covariant, right now the most used one is tuple type, and in addition to
+ * the control bit, tuple type's data bit is the 8th bit from the LSB. Control
+ * bits start from MSB and data bits start from LSB.
+ *
+ * If two types are equal, then they are subtype of each other, also if the bits
+ * from one type tag is subset of the other tag, it automatically becomes a
+ * subtype of the other. This simplifies the subtyping logic a lot, and over the
+ * long term it is possible to adopt this scheme on the server side as well.
+ * Special cases can be added but they generally should not take too much code
+ * size.
+ *
+ * DynamicType doesn't necessarily inherit from c10::Type, we might want to
+ * inherit from c10::Type to reduce the migration cost.
+ */
+class DynamicType : public Type {
```

Comment:
I mean "don't necessarily" by we may or may not inherit from c10::Type, and it's not the core requirement of DynamicType, but then follows "we might want to inherit from c10::Type to reduce the migration cost." to indicate there's practical consideration for the inehritance. 
I don't want to write things like "for now we do this, and in the long term we do that" because that could be more confusing for future readers.
I'll make it more clear.